chore: review Curation API spec

danieljhegeman opened the pull request:
### Reviewers
**Functional:** 
@Bento007 
**Readability:** 
@brianraymor @metakuni 
---

## Changes
- add
- remove
- modify

## Definition of Done (from ticket)

## QA steps (optional)

## Known Issues

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+        - $ref: "#/components/parameters/path_collection_uuid"
+        - $ref: "#/components/parameters/optional_query_curator_tag"
+        - $ref: "#/components/parameters/optional_query_dataset_uuid"
+        - $ref: "#/components/parameters/query_visibility"
```
Must be able to specify visibility so that they can both
1) delete a dataset from an ongoing revision `?visibility=private`
and
2) start a revision, and delete the dataset from the revision `?visibility=public`

danieljhegeman replied:
Alternatively we can outlaw #2 and force them to create a revision via PUT to `/curation/collections/{collection_id}/`

danieljhegeman replied:
forbid deletion of dataset from public collection

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+
+  /curation/collections/{collection_id}:
+    delete:
+      description: Delete a private Collection. Delete an existing revision for a public Collection.
```
Note: using this endpoint to delete an existing revision for a public Collection is a previously-undiscussed feature. I am not set on having it but I think it is a logical usage of the endpoint.

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+        - $ref: "#/components/parameters/path_collection_uuid"
+        - $ref: "#/components/parameters/optional_query_curator_tag"
+        - $ref: "#/components/parameters/optional_query_dataset_uuid"
+        - $ref: "#/components/parameters/query_visibility"
```
So the curator can fetch a private (revision) version of a public dataset.

Bento007 replied:
Aren't these all returned when they do `GET /collections/{collection_id}?visibility=private`? I think this endpoint is unnecessary.

danieljhegeman replied:
I agree it isn't necessary, just adds the ability to filter before launching the request. May be nice if a collection has hundreds of datasets, lower response time.

brianraymor replied:
We do not have collections with _hundreds of datasets_. The largest has 41, but most are much more limited.

danieljhegeman replied:
not needed

Bento007 commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+          description: OK
+          content:
+            application/json:
+              $ref: "#/components/schemas/collection"
```
I think here would be a good place to return both the public and the private in the same response. Similar to how you wanted to return the GET /collections endpoint.

danieljhegeman replied:
Since it's only two, sure. I agree!

Bento007 commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+          application/json:
+            $ref: "#/components/schemas/collection_metadata"
+      responses:
+        "204":
```
why 204?

danieljhegeman replied:
200 if existing (ok), 201 if new revision (created)

Bento007 commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+        - cxguserCookie: []
+      parameters:
+        - $ref: "#/components/parameters/path_collection_uuid"
+        - $ref: "#/components/parameters/optional_query_curator_tag"
```
for the `/curation/collections/{collection_id}/datasets` endpoint, we should put curator_tag and dataset_uuid in the request body rather than as a query parameter.

danieljhegeman replied:
why is that? To allow for batch deletes?

Bento007 commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+        - $ref: "#/components/parameters/optional_query_dataset_uuid"
+        - $ref: "#/components/parameters/query_visibility"
+      responses:
+        "204":
```
I think a 202 response is more appropriate.

Bento007 commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+          $ref: "#/components/responses/204"
+        "401":
+          $ref: "#/components/responses/401"
+        "404":
```
Remove this. We always return 202 on delete whether the file is there or not.

danieljhegeman replied:
What about for the super curator, who has permission to view everything?

danieljhegeman replied:
404

Bento007 commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+                  $ref: "#/components/schemas/dataset_uuid"
+                link:
+                  $ref: "#/components/schemas/upload_link"
+      responses:
```
need to add 400 response for when the link does not pass validation.

Bento007 commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+                link:
+                  $ref: "#/components/schemas/upload_link"
+      responses:
+        "204":
```
return 202 since this an async operation

danieljhegeman replied:
I know I assented to this when we were meeting earlier today, but I disagree after thinking about it more. The actions taken by the server have been completed; there is no queued work. The fact that the next thing for the curator to do is to upload to s3 does not impinge on whether the response from this endpoint is considered to be asynchronous at all.

Bento007 commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+            application/json:
+              schema:
+                type: object
+                additionalProperties: false
```
```suggestion
```

danieljhegeman replied:
not a valid response body attribute

Bento007 commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+                        required: true
+        "401":
+          $ref: "#/components/responses/401"
+        "404":
```
when would we return a 404?

danieljhegeman replied:
will delete to reduce complexity

danieljhegeman replied:
So you agreed that we could leave the 404 for `/curation/collections/{collection_id}/datasets`, which sort of implies we might want to leave it for this route as well _for super curators only_.

What do you think?

danieljhegeman replied:
Hm. I definitely don't agree that we couldn't possibly need a 404 for _any_ reason (see my messages immediately above; if the collection doesn't exist and the curator is a super-user). 

But I would agree it isn't an important part of the response outline.

danieljhegeman replied:
removed (`51f8fa7f`)

Bento007 commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+        email:
+          description: The user's email
+          type: string
+    collection_metadata:
```
Remove published_at and revised_at. These are not fields the user has permission to modify. Make a separate schema for the collection metadata response if it will include the additional fields. Be careful when reusing schemas. The schema conveys meaning in the backend and could result in us giving the users more privilege than we intend.

danieljhegeman replied:
Great point. The schemas need to be different for request and response.

Bento007 reviewed (changes requested):
Great work overall. There are some areas that need refining.
- Remove the schemas and components that are never used. 
- You don't need `required` if the schema is only used in a response.
- add `additional_field: false` to all requests that contain a body.

Bento007 commented:
To make this easier to review I suggest putting in swagger up and sharing a link to the API.

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+      security:
+        - cxguserCookie: []
+      responses:
+        "202":
```
Because a 200 indicates the key was found, whereas 202 indicates we'll delete it *if* it exists?

Bento007 replied:
exactly

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+                  collections:
+                    type: object
+                    properties:
+                      public:
```
Another question is whether we want to return public collections with their private revision _as a pair_ if the user wants both.

danieljhegeman replied:
related: https://czi-sci.slack.com/archives/C02UGUV5R3J/p1645208034691689?thread_ts=1645146245.806889&cid=C02UGUV5R3J

Bento007 replied:
>Are you saying you don't like the idea of collections being separated into different arrays by visibility?

Yes. It's an extra step we need to do on the backend, that can be easily accomplished by the curator on their end. This also becomes complicated as the number of collections grows. We will not be able to return all of the collections at once and will need to paginate.

danieljhegeman replied:
Will return single array of collections. Private revisions get public id's.

brianraymor replied:
> Yes. It's an extra step we need to do on the backend, that can be easily accomplished by the curator on their end.

If it's "easily accomplished" for the curator, then the same is true for the BE.

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+      responses:
+        "201":
+          description: Created
+            content:
```
I agree with this one 👍

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+                  $ref: "#/components/schemas/collection_uuid"
+        "401":
+          $ref: "#/components/responses/401"
+
```
Good catch

danieljhegeman replied:
Added 400; will use the existing response schema for a single missing or invalid parameter

danieljhegeman replied:
@Bento007 @brianraymor wanted to draw your attention to this issue again: I think it seems reasonable to use the existing 404 structure to return an error for the first invalid or missing parameter that we encounter.

OR would you prefer our backend to assess all parameters and return a list of invalid/missing parameters?

brianraymor replied:
I prefer:

> backend to assess all parameters and return a list of invalid/missing parameters? 

we also want to reuse this pattern/code when we transition validation of the collection information in the UX to the BE which is a fit&finish issue for _filter by metadata_.

danieljhegeman replied:
That format is fine. I was simply referencing the existing 404 format. I can make the change to cover _all_ parameters on a single call.

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+      parameters:
+        - $ref: "#/components/parameters/path_collection_uuid"
+      responses:
+        "200":
```
the rest of the response would just be the `description`, right?

danieljhegeman replied:
make the change dan

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+          $ref: "#/components/responses/401"
+        "404":
+          $ref: "#/components/responses/404"
+
```
I'll add the 400 👍

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,829 @@
+                      status:
+                        type: string
+                        enum: [FAILURE, PENDING, SUCCESS]
+                        required: true
```
remove

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+                properties:
+                  collections:
+                    type: array
+                      items:
```
When I copy the raw file into [swagger.io](https://editor.swagger.io/), line 84 results in:
```
Parser error 
bad indentation of a mapping entry
```

danieljhegeman replied:
I'll proof this 👍

brianraymor commented:
General comment - it would be really helpful for reviewers if there were more `description`(s) in the swagger.

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+                              type: string
+                            data_submission_policy_version:
+                              type: string
+                            genesets:
```
`genesets` should be removed. This is leftover from an earlier design that was deprecated.

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+                              type: number
+                            description:
+                              type: string
+                            obfuscated_uuid:
```
What is `obfuscated_uuid` ?

danieljhegeman replied:
Ok. Had maintained this from the other API, but since it's not needed for (presumably-hidden) FE logic, will remove.

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+                            revision_of:
+                              type: string
+                              description: uuid of the public version of this Collection
+                            access_type:
```
What is `access_type`? Does this indicate whether I own the collection and have write access?

Bento007 commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+      security:
+        - cxguserCookie: []
+      responses:
+        "200":
```
add 404 if the user does not have a key created.

brianraymor commented:
General comment - there should be an explicit description that the API(s) related to the the curator API key are called from the portal and not from curator code. I guess that:

```
      security:
        - cxguserCookie: []
```

is an implicit indication.

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+        "401":
+          $ref: "#/components/responses/401"
+    get:
+      summary: Fetch the identifier of the curator's active API keys
```
Is there more than one key? - " active API **keys**".  And this is a portal internal API and not a developer API.

danieljhegeman replied:
Thanks. We're going to restrict to a single active key. Will change this text

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+        "401":
+          $ref: "#/components/responses/401"
+    delete:
+      summary: Delete the curator's API key
```
This is an internal portal API.

Bento007 replied:
This and all of the key and token APIs will be mainly portal specific, but they still need to be defined somewhere. We can move them to the Other API later.

brianraymor replied:
I think if we just flag it as "internal" in the description, then there's more awareness for reviewers until we can offer a demonstration notebook with step by step instructions?

danieljhegeman replied:
Ok, will add flag

metakuni replied:
Maybe we could use tags for now:
- https://swagger.io/docs/specification/grouping-operations-with-tags/

Bento007 replied:
Tag could be in addition, but we should put internal in the Summary and or Description.

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+
+  /curation/collections:
+    get:
+      description: Fetch all Collections metadata. Authentication required for private collections.
```
Would it not be authorization rather than authentication?

danieljhegeman replied:
Correct! 👍

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+                              type: string
+                            visibility:
+                              $ref: "#/components/schemas/visibility"
+                            contact_name:
```
General question: For names of humans (e.g. `contact_name`, `curator_name`) do we want/expect `<first name> <last name>` or `<last name>, <first name>`? Should this be documented in a description somewhere?

brianraymor replied:
We're _preferring_ `<first name> <last name>` for the contact, but obviously cannot enforce it. I've seen a recent collection where the contact is an info mailing list for Kidney Precision Medicine Project (KPMP) - see https://cellxgene.cziscience.com/collections/bcb61471-2a44-4d00-a0af-ff085512674c.

The `curator_name` is _automagically_ set by the BE using Auth0.

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+
+  /curation/auth/token:
+    post:
+      summary: Generate an access token
```
Nits:

descripton -> description
to the use -> to the user

Perhaps:

Returns a bearer access token that must be passed as a parameter to requests that require authorization such as ...

Bento007 replied:
> Returns a bearer access token that must be passed as a parameter to requests that require authorization such as
:+1

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+                            - contact_email
+                            - curator_name
+                            - revised_at
+                            - created_at
```
This is missing `published_at`. 

`created_at` is internal to the portal and indicates when the table was updated - which can be a placeholder for when a private collection was created.

brianraymor replied:
I see the property for missing `published_at` is defined below.

danieljhegeman replied:
Just to confirm: you're still ok with returning `created_at` as well, right?

brianraymor replied:
I'm fine with returning `created_at` because it allows @jahilton to determine how long a private collection has been _baking_ prior to publication.

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+                  - contact_name
+                  - contact_email
+                  - curator_name
+                properties:
```
Should this be replaced with `allOf` and  `$ref: "#/components/schemas/query_collection_metadata"` and the extra properties `published_at`, `revised_at` and `datasets`?

danieljhegeman replied:
In actuality, no. Because the collection data we take in for POST/PUT of a collection is not de facto a logical subset of the stuff we would return for a GET.

reverted (`7ab414f3`)

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+                    properties:
+                      status:
+                        type: string
+                        enum: [FAILURE, PENDING, SUCCESS]
```
We're not exposing detailed statuses for upload and conversions?

Might an API user not want to know if say, failures caused by a Seurat conversion vs an inaccessible upload URL? (for instance)

ebezzi replied:
I think that's defined by the `type` parameter above. I am pretty sure the current endpoint returns detailed information about each single conversion (cxg, rds etc)

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+      parameters:
+        - $ref: "#/components/parameters/path_collection_uuid"
+        - $ref: "#/components/parameters/optional_query_curator_tag"
+        - $ref: "#components/parameters/optional_query_dataset_uuid"
```
If the dataset UUID is an optional parameter, the response should probably be an array with one status per dataset.

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+    user_uuid:
+      description: A unique identifier of a logged in User of Corpora.
+      type: string
+    userinfo:
```
Unused.

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+            type: string
+          link_type:
+            type: string
+            enum: [PROTOCOL, RAW_DATA, DOI, LAB_WEBSITE, OTHER, DATA_SOURCE]
```
Should we add a description that defines each of these enum values?

danieljhegeman replied:
[3.1.0](https://stackoverflow.com/questions/66465888/how-to-define-enum-mapping-in-openapi) makes this easy! We'll do it.

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+          type: array
+          items:
+            $ref: "#/components/schemas/dataset_asset"
+        tombstone:
```
Should tombstone status be exposed to the API user?

danieljhegeman replied:
Yes. Jason wants to see deleted datasets as well.

brianraymor commented:
I've asked @ebezzi to review because we need to align curation API and filter by metadata API changes. One example is that there is now publication metadata included for collections that define a DOI for _filter by metadata_, which is missing from the curation API.

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+        linked_genesets:
+          type: array
+          items:
+            type: string
```
Is this an array of gene set UUIDs?

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+          type: integer
+        X_approximate_distribution:
+          $ref: "#/components/schemas/distribution"
+        schema_version:
```
Should `schema_version` be extracted as a data type and defined as an enum?

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+      summary: Generate the curator's API key
+      security:
+        - cxguserCookie: []
+      responses:
```
That's not a 403? 🤔

brianraymor replied:
Is there a reason not to regenerate the API key?

Otherwise ... 

It does not seem to be a 409 based on my reading - https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.8 - how about 400 - https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.1

Bento007 replied:
based on that [link](https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.8) 409 seems incorrect, but 400 also seems wrong. I believe we settle on regeneration of the API key.

danieljhegeman replied:
I'm fine with that. It will regenerate.

danieljhegeman replied:
Updated description `90ddef70`

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+          $ref: "#/components/schemas/distribution"
+        schema_version:
+          type: string
+    dataset_preview:
```
Should we combine the overlap between `dataset` and `dataset_preview` by using `allOf` and `$ref`? (to avoid repetition)

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+      properties:
+        name:
+          type: string
+        id":
```
```suggestion
        id:
```

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+          type: string
+        label:
+          type: string
+    ontology_element_array:
```
To follow the general naming convention, should this be `ontology_elements`?

danieljhegeman replied:
It's not entirely clear to me that there's precedent in our spec for pluralizing the descriptor whenever we append `_array`. This is the only instance where we do that.

Bento007 replied:
In that case I second `ontology_elements`

danieljhegeman replied:
Oh I totally misinterpreted the first time around. I was thinking you were both suggesting `ontology_elements_array`. Will change.

danieljhegeman replied:
fixed (`d2586b4d`)

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+      description: Error message container for HTTP APIs.
+      properties:
+        type:
+          type: string
```
Should this be an enum?

danieljhegeman replied:
It could be, but it doesn't really matter too much. Custom descriptor from the BE.

metakuni reviewed (commented):
I've only _just_ read the OpenAPI spec for the first time, so apologies in advance for newb comments,  questions, and pedantry. 😄

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/query_collection_metadata"
```
Query is not the right term, you're right. I'll replace it with `request_collection_metadata` or `input_collection_metadata` or something like that. Maybe `collection_form_metadata`.

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+
+  /curation/collections/{collection_id}/datasets/upload-links:
+    put:
+      description: Upload a Dataset. The Dataset will be downloaded from the provided link.
```
I meant downloaded by us. Uploaded/downloaded... but I'll change it.

danieljhegeman replied:
fixed `28326bc3`

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+                              type: string
+                            obfuscated_uuid:
+                              type: string
+                            data_submission_policy_version:
```
`data_submission_policy_version` is not particularly useful for curators. It's only set when the collection is published or revised/published when the curator accepts the data submission policy on behalf of the data submitter. 

It's only updated when the policy is updated - which has happened only once. And when it's updated, there's an out-of-band method for getting permission from all the curators to update the state in legacy collections via a script.

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+          enum: [H5AD, RDS, CXG]
+        filename:
+          type: string
+        type:
```
`type` should be deprecated. The original model for the portal was to offer the original dataset from the data submitter plus the cellxgene "remix" which re-mapped the original dataset to the cellxgene schema.

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+          type: string
+        filetype:
+          type: string
+          enum: [H5AD, RDS, CXG]
```
@jahilton - would you validate that the CXG asset was created through the API or use the UI to explore the dataset to validate?

brianraymor replied:
And this is a case where the permanent link is actually in datasets_deployment.

danieljhegeman replied:
@brianraymor can you please clarify? Jason will need to validate whether the the asset was created via an API submission vs UI submission?

brianraymor replied:
I do not see that the CXG asset in `dataset_assets` is valuable to curators, so I would suggest that it is removed:

"dataset_assets": [
          {
            "created_at": 1638224578.183704,
            "dataset_id": "524e045e-e74c-4e00-9884-d5c3bef3d862",
            "filename": "explorer_cxg",
            "filetype": "CXG",
            "id": "e8b0e0ce-41aa-4666-ae2c-c10098c98cc6",
            "s3_uri": "s3://hosted-cellxgene-prod/524e045e-e74c-4e00-9884-d5c3bef3d862.cxg/",
            "type": "REMIX",
            "updated_at": 1638224578.183711,
            "user_submitted": true
          },

On the other hand, the permanent link in `dataset_deployments` can be both used to validate that the CXG was created (from the API) and to capture the link to share with data submitters if needed:

        "dataset_deployments": [
          {
            "url": "https://cellxgene.cziscience.com/e/524e045e-e74c-4e00-9884-d5c3bef3d862.cxg/"
          }
        ],

danieljhegeman replied:
Ok. Keep permanent link to cxg but remove from dataset_assets (leaving just H5AD and RDS).

jahilton replied:
>would you validate that the CXG asset was created through the API or use the UI to explore the dataset to validate?

I would use the API.
>I do not see that the CXG asset in dataset_assets is valuable to curators, so I would suggest that it is removed:

I agree

danieljhegeman replied:
removed CXG enum (`3b8cc6a4`)

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+          type: string
+        type:
+          type: string
+          enum: [REMIX, ORIGINAL]
```
type should be deprecated. The original model for the portal was to offer the original dataset from the data submitter plus the cellxgene "remix" which re-mapped the original dataset to the cellxgene schema.

danieljhegeman replied:
... so it's no longer used, got it. Removed. 👍

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+          enum: [REMIX, ORIGINAL]
+        s3_uri:
+          type: string
+    gene_set:
```
And all need to be deprecated per my general comment.

danieljhegeman replied:
Removed in `061be86f`

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+openapi: 3.0.0
```
Good question. TBD

danieljhegeman replied:
@Bento007 let me know your thoughts here. I'm fine with moving to 3.1.0. Doesn't look like [the breaking changes](https://github.com/OAI/OpenAPI-Specification/releases/tag/3.1.0-rc1) will affect us.

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,725 @@
+      parameters:
+        - $ref: "#/components/parameters/path_collection_uuid"
+        - $ref: "#/components/parameters/optional_query_curator_tag"
+        - $ref: "#components/parameters/optional_query_dataset_uuid"
```
```suggestion
        - $ref: "#/components/parameters/optional_query_dataset_uuid"
```

metakuni commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,725 @@
+      parameters:
+        - $ref: "#/components/parameters/path_collection_uuid"
+        - $ref: "#/components/parameters/optional_query_curator_tag"
+        - $ref: "#components/parameters/optional_query_dataset_uuid"
```
```suggestion
        - $ref: "#/components/parameters/optional_query_dataset_uuid"
```

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+        type:
+          type: string
+          enum: [REMIX, ORIGINAL]
+        s3_uri:
```
How would a curator use the `s3_uri`?

Bento007 replied:
they don't really have a use for it.

danieljhegeman replied:
removed (`7b42f47d`)

Had left that in when I initially thought that those were already presigned urls. But since they aren't, I added in the `/curation/collections/{collection_uuid}/asset` route yesterday.

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,729 @@
+          $ref: "#/components/schemas/ontology_element_array"
+        development_stage:
+          $ref: "#/components/schemas/ontology_element_array"
+        dataset_deployments:
```
I've never understood why `dataset_deployments` is an array since it simply contains the url for the CXG permanent link:

```
        "dataset_deployments": [
          {
            "url": "https://cellxgene.cziscience.com/e/aad97cb5-f375-45ef-ae9d-178e7f5d5180.cxg/"
          }
```

brianraymor replied:
I would recommend the following changes:

1. This does not need to be an array.
2. The name is not self-documenting. What is a "dataset deployment" to a curator. @jahilton @ambrosejcarr - any suggestions for a rename related to visualization, explorer, cxg, etc?

jahilton replied:
`visualization_url` or `explorer_url`

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,729 @@
+      properties:
+        id:
+          type: string
+        dataset_id:
```
In the current portal API, the `dataset_id` is repeated in each `dataset_asset` and is also available in datasets `id`.  That seems redundant.

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,729 @@
+        - curator_tag
+        - name
+        - revision
+        - collection_id
```
Is there a need to duplicate the `collection_id` in the dataset? (This currently occurs in the portal API)

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+              schema:
+                type: object
+                additionalProperties: false
+                required:
```
Can you point me to where you're finding that `required` should never be in a Response?

Same is true for `additionalProperties` too, though, right?

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+          type: string
+        detail:
+          type: string
+    query_collection_metadata:
```
To indicate it is the metadata associated with a request from the user. It was a poor naming choice due to swapping things in and out a couple times. Changing `collection_form_metadata`.

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+      in: cookie
+      name: cxguser
+      x-apikeyInfoFunc: backend.corpora.lambdas.api.v1.authentication.apikey_info_func
+    cxguserCookieLenient:
```
Don't we want to allow anyone to retrieve public collections from `/collections`?

ebezzi replied:
This depends on whether we want to use these APIs for the UI as well. The `lenient` function was introduced to ignore all possible authentication errors on endpoints that should be 100% public (for instance, these can occur if some bad authentication token is stored in the cookies). This is useless in an API context, but if we plan to reuse these APIs, we should keep it, or at least incorporate a more lenient logic into the main authentication mechanism.

Bento007 replied:
Do we have an existing test that uses the lenient cookie?

danieljhegeman replied:
I feel like we're talking to different issues. I don't see why we would lock down the `/collections` endpoint.

danieljhegeman replied:
@Bento007 ah, the purpose of the lenient cookie is to allow tracking of browser metadata without using an auth mechanism. 

Which we wouldn't need for the API. Ok. 

(that's the right reasoning, yes?)

brianraymor replied:
> Don't we want to allow everybody, including those without API keys, to retrieve public collections from /collections?

Yes we do. The minimal implementation will allow anyone to query public collections/datasets and to download datasets in the format of their choice.

danieljhegeman replied:
@Bento007 bump

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+      in: cookie
+      name: cxguser
+      x-apikeyInfoFunc: backend.corpora.lambdas.api.v1.authentication.apikey_info_func_lenient
+    dummyAuth:
```
they aren't 😎

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+        tombstoned:
+          type: boolean
+    upload_link:
+      description: A link from which to download a Dataset source file.
```
updated, let me know how you feel

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+            $ref: "#/components/schemas/dataset_asset"
+        tombstone:
+          type: boolean
+        assay:
```
We go with what the column names are in the larger cellxgene product. Perhaps a cleanup issue for a later date.

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+        id:
+          type: string
+        dataset_id:
+          type: string
```
Removing

danieljhegeman replied:
I think we want to leave this as-is. Needs to stay as `id`. And we don't use a `dataset_asset_uuid` anywhere else.

ebezzi commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,683 @@
+        "401":
+          $ref: "#/components/responses/401"
+
+  /curation/collections:
```
Consider whether this endpoint is needed. This is effectively a dump of all of our database, and will scale poorly in the future (even now, that's a few megabytes of data). For comparison, the FE only receives a subset of this data on load.

An alternative is to add pagination.

danieljhegeman replied:
Yep, it'll be a good thing to keep tabs on as we move forward.

For now, it is desired for our curators. We'll see about adding pagination if it's absolutely necessary or if we have a little spare time at the end.

ebezzi commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,683 @@
+      required: true
+      schema:
+        $ref: "#/components/schemas/collection_uuid"
+    query_user_uuid:
```
If this refers to links to our Auth0 users, those aren't uuids.

danieljhegeman replied:
removed (`4b57acc4`)

ebezzi commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,683 @@
+        $ref: "#/components/schemas/visibility"
+        default: PUBLIC
+
+  responses:
```
Seems a bit weird that we should provide descriptions here, as these are standard HTTP codes.

ebezzi commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,683 @@
+          content:
+            application/json:
+              schema:
+                allOf:
```
This seems wrong. This should return the same response as the `GET /curation/collections` but for a singe collection.

danieljhegeman replied:
Just using some inheritance concepts. The `collection_form_metadata` is the subset of collection data that is user-provided. The stuff returned from this collection endpoint is 1) all of that, plus 2) the fields that get added (published_at, revised_at, datasets).

danieljhegeman replied:
Although I suppose that, since the submission form does not necessarily mean we would want to return that field for the collections endpoint, perhaps I should split this back out. Just because it happens to line up doesn't mean it makes logical sense. I assume that's your criticism.

ebezzi commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,683 @@
+                    type: array
+                    items:
+                      type: object
+                      properties:
```
`publisher_metadata` is missing. My bad for not including it in the current API spec, I'll do it and add it here as a suggestion.

danieljhegeman replied:
added per your PR

ebezzi commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,683 @@
+                          type: number
+                        description:
+                          type: string
+                        obfuscated_uuid:
```
What is this field for again?

danieljhegeman replied:
need to remove

danieljhegeman replied:
Removed in `061be86f`

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+    dataset_uuid:
+      description: A unique identifier of a Dataset.
+      type: string
+    geneset_uuid:
```
addressed in `39db4d3d`

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+  schemas:
+    user_uuid:
+      description: A unique identifier of a logged in User of Corpora.
+      type: string
```
Good question -- going to tag @Bento007 to ask whether this was considered for the previous api

metakuni replied:
On reflection, perhaps this doesn't matter since UUIDs are always generated by us, not the API user.

Bento007 replied:
let's skip the regex since they are all generated by us.

Bento007 commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+        name:
+          description: The user's name
+          type: string
+        email:
```
I'm not sure how connexion will resolve this. It's worth trying.

Bento007 commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,683 @@
+      parameters:
+        - $ref: "#/components/parameters/path_collection_uuid"
+      responses:
+        "202":
```
it's async because the s3 assets may not be removed immediately.

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+
+components:
+  schemas:
+    user_uuid:
```
No, need to have them separate for cases where they are used as identifiers within the spec.

danieljhegeman replied:
Actually, I was under the misimpression that the unique names would be more useful for parts of the auto-generated docs than they are. I think we can consolidate as you suggested.

danieljhegeman replied:
done (`e640156c`)

brianraymor commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,721 @@
+    dataset_asset:
+      type: object
+      properties:
+        id:
```
The `id` of a `dataset_asset` is a duplicate of the `dataset_id` (called `id` above) of a dataset in the current portal API. It should be removed.

For example:

```
Dataset: Single-nucleus RNA-seq of the Adult Human Kidney (Version 1.0) uuid: 07854d9c-5375-4a9b-ac34-fa919d3c3686
asset[0] uuid - 07854d9c-5375-4a9b-ac34-fa919d3c3686
asset[1] uuid - 07854d9c-5375-4a9b-ac34-fa919d3c3686
asset[2] uuid - 07854d9c-5375-4a9b-ac34-fa919d3c3686

```

danieljhegeman replied:
Ah, got it, I didn't realize that. Thanks Brian!

danieljhegeman replied:
resolved (`96d5318c`)

brianraymor replied:
I think it would be helpful to always use a specific name for an id rather than relying on context.

Clarifying my point about duplication of uuids on the call. In the current portal API, each dataset in `datasets` defines an `id` which is the dataset id. This id is then duplicated in each asset in `dataset_assets` which includes the dataset id as `dataset_id` and the asset id as `id'.  We should avoid such duplication in this API. 


        "dataset_assets": [
          {
            "created_at": 1638224578.183704,
            "dataset_id": "524e045e-e74c-4e00-9884-d5c3bef3d862",
            "filename": "explorer_cxg",
            "filetype": "CXG",
            "id": "e8b0e0ce-41aa-4666-ae2c-c10098c98cc6",
            "s3_uri": "s3://hosted-cellxgene-prod/524e045e-e74c-4e00-9884-d5c3bef3d862.cxg/",
            "type": "REMIX",
            "updated_at": 1638224578.183711,
            "user_submitted": true
          },

        "id": "524e045e-e74c-4e00-9884-d5c3bef3d862",

---

In the current portal API, there is also a duplication of the `collection_id` between the collection level and the dataset level.

Bento007 reviewed (approved):
This is enough of a spec for us to get started. I'm sure some things will change once we start implementing. Before committing move the spec to backend/config.

brianraymor commented:
`mean_genes_per_cell` should be included with datasets - for download scenarios rather than for curators.

brianraymor commented:
Is there a tombstone attribute for collections in addition to datasets? We have deleted public collections.

brianraymor commented:
The `curator_name` MUST NOT be included in `POST curation/collections` because it is set by the BE and NOT the curator to ensure consistency:

{
  "name": "string",
  "description": "string",
  "contact_name": "string",
  "contact_email": "string",
  "curator_name": "string",
  "links": [
    {
      "link_name": "string",
      "link_url": "string",
      "link_type": "PROTOCOL"
    }
  ]
}

brianraymor commented:
RE `data_submission_policy_version` for a collection - while I would want this returned by "some API" to allow me to monitor migration of collections to the next version of the data submission policy (which has happened once), it's unclear whether it's useful to curators since we don't simultaneously support multiple versions of the policy. It's mainly for internal auditing. 

@jahilton - any opinions?

jahilton commented:
>RE `data_submission_policy_version` for a collection...it's unclear whether it's useful to curators

I would not find this useful.

brianraymor commented:
@jahilton and I propose an addition. Similar to the CXG URL, we should add a collection field for the `collection_url`.

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,762 @@
+                  data_submission_policy_version:
+                    type: string
+                  publisher_metadata:
+                    $ref: "#/components/schemas/publisher_metadata"
```
added the publisher_metadata

danieljhegeman replied:
Also, there were columns that were returned from `/curation/collections` which were not being returned here; I've added them to be consistent. Let me know if you disagree @brianraymor @Bento007

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,762 @@
+                        data_submission_policy_version:
+                          type: string
+                        publisher_metadata:
+                          $ref: "#/components/schemas/publisher_metadata"
```
added publisher metadata

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,762 @@
+        Determines the visibility of the collection, that will either make the collection visible on the public
+        sites or only viewable through obfuscated URLs.
+      type: string
+      enum: [PUBLIC, PRIVATE, REVISION]
```
added (`a71ebf9d`)

cc @jahilton @Bento007 @brianraymor

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,790 @@
+openapi: 3.0.0
+info:
```
added version and title (`e7e4f514`)

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,753 @@
+        - $ref: "#/components/parameters/path_collection_uuid"
+        - $ref: "#/components/parameters/query_curator_tag"
+        - $ref: "#/components/parameters/query_dataset_uuid"
+        - $ref: "#/components/parameters/query_visibility"
```
Added visibility so that the curator can specify which dataset they want to download.

brianraymor commented:
How does a curator update collection information in private collections or revisions - similar to `... Edit Details` in the UX?

danieljhegeman commented:
> How does a curator update collection information in private collections or revisions - similar to `... Edit Details` in the UX?
> 
> It appears that a curator can use [PUT /curation/collections/collections_uuid](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/put_curation_collections__collection_uuid_) to start or update a revision, but how is the collection information in a private collection updated prior to publication?

The private collection is identified using the private collection id (or, if it's a revision, then the public collection id):

`PUT` to `/curation/collections/{collection_id}`. If the collection is private, it is updated in place. If the collection is a revision, it is updated in place. If the collection is public and does not have a revision started, a revision is started and the fields are updated for the revision.

brianraymor commented:
> If the collection is public and does not have a revision started, a revision is started and the fields are updated for the revision.

I thought we agreed earlier that no magic would occur. This would fail.

brianraymor commented:
How does [POST /curation/auth/token](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/post_curation_auth_token) return a token when no parameters are included in the request?

danieljhegeman commented:
> > If the collection is public and does not have a revision started, a revision is started and the fields are updated for the revision.
> 
> I thought we agreed earlier that no magic would occur. This would fail.

Can you say more? I think the lines around "magic" are blurry. What specific pattern are we using here that is worrying?

brianraymor commented:
I don't think we should start a revision because it was implicitly suggested by the developer performing an update to a public collection. The API should be minimal and not make assumptions. 

Curator - Dang, I made a mistake and now I need to go and delete a revision!

brianraymor commented:
RE [GET /curation/collections](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/get_curation_collections)

@jahilton - would you expect that the `visibility` parameter would allow you to query for multiple "visibilities" - PUBLIC, PRIVATE and/or REVISION? Current, only one can be passed.

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,757 @@
+          name: Authorization
+          schema:
+            type: string
+          required: true
```
^^ @Bento007 

cc @brianraymor

danieljhegeman replied:
If this is the right format, I'll add to the other endpoints as well (for the token)

brianraymor replied:
Authorization would not be **required** for other endpoints if only public access was desired?

danieljhegeman commented:
> I don't think we should start a revision because it was implicitly suggested by the developer performing an update to a public collection. The API should be minimal and not make assumptions.
> 
> Curator - Dang, I made a mistake and now I need to go and delete a revision!

This is a very mild implicit step IMO. Furthermore, if a revision exists already, then this is the request pattern they would use to edit the revision. I would bucket the act of forgetting that an old revision exists and accidentally overwriting it in with making the mistake of starting a revision. And we can't prevent all mistakes. I think that if we were to extend that logic we'd want to get rid of the DELETE endpoint for private collections as well. 

Thoughts @metakuni @Bento007 ?

brianraymor commented:
RE [GET /curation/collections](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/get_curation_collections)

How are tombstoned collections modeled? (Only datasets have a `tombstone`) Is this a fourth `visibility` ?

brianraymor commented:
RE RE [GET /curation/collections] (https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/get_curation_collections) is described as:

 _Fetch **all** Collections metadata. Authorization required for private collections._ [and revisions of public collections].

Please update the description to document that a limited preview of metadata for datasets is returned in the response. Full dataset metadata is available from [GET /curation/collections/{collection_uuid](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/get_curation_collections__collection_uuid_)

brianraymor commented:
RE [/curation/collections/{collection_uuid}](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/get_curation_collections__collection_uuid_)

The documentation states:

_Fetch Collection metadata and associated Datasets, public and private._

This could simply be _Returns complete collection metadata for the collection identified by uuid._

uuid - _The uuid of a collection. To specify the private revision of a public collection, use the uuid of the public collection._
visibility - _Available values : PUBLIC, PRIVATE, REVISION_

I don't think that the visibility enum makes sense in this case.

1. If a curator wants to query a public collection, the public uuid is passed. No visibility is required.
2. If a curators wants to query a private collection, the private uuid is passed. No visibility is required.
3. If a curator wants to query a revision, the public uuid is passed AND there needs to be an indication that the REVISION is desired. So perhaps we simply need a ?revision=true query?

brianraymor commented:
RE [/curation/collections/{collection_uuid}](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/get_curation_collections__collection_uuid_)

Why does `datasets` duplicate information available in the collection metadata?

      "collection_id": "string",
      "collection_visibility": "PUBLIC",

brianraymor commented:
RE [/curation/collections/{collection_uuid}](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/get_curation_collections__collection_uuid_)

This earlier [comment](https://github.com/chanzuckerberg/single-cell-data-portal/pull/1982/#discussion_r814294895) has not been addressed to rename and retype:




      "dataset_deployments": [
        {
          "url": "string"
        }
      ],

brianraymor commented:
RE [/curation/collections/{collection_uuid}](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/get_curation_collections__collection_uuid_)

`mean_genes_per_cell` is missing from `datasets` for each dataset.

brianraymor commented:
RE [/curation/collections/{collection_uuid}](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/get_curation_collections__collection_uuid_)

`published_at` is missing for each dataset in `datasets`.

brianraymor commented:
RE [POST /curation/collections](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/post_curation_collections)

The curator is not allowed to set the `curator_name` when creating a collection. For consistency, it is set by the BE using Auth0. 


{
  "name": "string",
  "description": "string",
  "contact_name": "string",
  "contact_email": "string",
  ~"curator_name": "string",~
  "links": [
    {
      "link_name": "string",
      "link_url": "string",
      "link_type": "PROTOCOL"
    }
  ]
}

brianraymor commented:
RE [POST /curation/collections](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/post_curation_collections)

Is there supposed to be more information in the swagger response related to the 201?

_The HTTP 201 Created success status response code indicates that the request has succeeded and has led to the creation of a resource. The new resource is effectively created before this response is sent back **and the new resource is returned in the body of the message,** its location being either the URL of the request, or the content of the [Location](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Location) header._

brianraymor commented:
RE [PUT /curation/collections/{collection_uuid}](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/put_curation_collections__collection_uuid_)

The curator is not allowed to set the `curator_name`. For consistency, it is set by the BE using Auth0 on collection creation.


{
  "name": "string",
  "description": "string",
  "contact_name": "string",
  "contact_email": "string",
  ~"curator_name": "string",~
  "links": [
    {
      "link_name": "string",
      "link_url": "string",
      "link_type": "PROTOCOL"
    }
  ]
}

brianraymor commented:
RE [PUT /curation/collections/{collection_uuid}](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/put_curation_collections__collection_uuid_)

The description states _Start a revision of a public Collection, or update its existing revision. Update a private Collection in-place. The Collection metadata fields included in the body of the request are updated._

1. I strongly recommend that a PUT to a public collection NOT initiate a revision, but return an error.2. 
1. The operation is a full replacement of the collection metadata and not an update (?); otherwise, how could a curator delete an existing link?

danieljhegeman commented:
@brianraymor Would I correct in surmising that you want specification of a revision to always require 'public_collection_id PLUS visibility = REVISION'?

brianraymor commented:
RE [DELETE /curation/collections/{collection_uuid}/datasets](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/delete_curation_collections__collection_uuid__datasets)

The documentation for the `curator_tag` and `dataset_uuid` parameters is unclear. Can you provide examples? If a curator does not pass either, what's the behavior?

jahilton commented:
> RE [GET /curation/collections](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/get_curation_collections)
> 
> @jahilton - would you expect that the `visibility` parameter would allow you to query for multiple "visibilities" - PUBLIC, PRIVATE and/or REVISION? Current, only one can be passed.

I would not expect to be able to query multiple visibilities. Only passing one is fine.

brianraymor commented:
RE [DELETE /curation/collections/{collection_uuid}](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/delete_curation_collections__collection_uuid_)

What is the return status if the `collection_uuid` is invalid - for example, the resource does not exist or a public collection uuid is specified, but there's no revision in progress.

brianraymor commented:
RE [GET /curation/collections/{collection_uuid}/asset](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/get_curation_collections__collection_uuid__asset)

I wish that we had a better term than asset.

Is `filetype` intentionally missing? 

For example, the current API includes with assets:

  "filetype": "RDS",

In addition, the documentation for the query parameters is unclear.  Can you provide examples?

brianraymor commented:
RE [PUT /curation/collections/{collection_uuid}/datasets/upload-link](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/put_curation_collections__collection_uuid__datasets_upload_link)

Is this used for both uploading a new dataset and replacing an existing dataset? If so, the request body must be clarified. For example, a curator would not define an `id` for a new dataset.

 

brianraymor commented:
RE [PUT /curation/collections/{collection_uuid}/datasets/upload-s3](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/put_curation_collections__collection_uuid__datasets_upload_s3)

There needs to be documentation for how the response is used.

I'm curious why this is a PUT and not a GET. I'm not keen on "upload-s3" since it returns credentials that can be used to upload to s3 using some other API.

brianraymor commented:
RE [GET /curation/collections/{collection_uuid}/datasets/status](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/get_curation_collections__collection_uuid__datasets_status)

How is a curator informed that an upload failed due to a schema validation error?

brianraymor commented:
How does a curator cancel an _in progress_ upload of a dataset? This is possible from the UX.

Can a curator delete a private collection or cancel a private revision of a public collection if an upload is _in progress_? Both are possible in the UX.

A curator cannot publish a collection with an _in progress_ upload because the UX will prevent that.

danieljhegeman commented:
> RE [/curation/collections/{collection_uuid}](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/get_curation_collections__collection_uuid_)
> 
> Why does `datasets` duplicate information available in the collection metadata?
> 
> ```
>   "collection_id": "string",
>   "collection_visibility": "PUBLIC",
> ```
> 
> **Update:** `id` is not defined in the collection metadata. It should be included.

Addressed in `d2e2b452`

danieljhegeman commented:
> RE [PUT /curation/collections/{collection_uuid}/datasets/upload-s3](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/put_curation_collections__collection_uuid__datasets_upload_s3)
> 
> There needs to be documentation for how the response is used.
> 
> I'm curious why this is a PUT and not a GET. I'm not keen on "upload-s3" since it returns credentials that can be used to upload to s3 using some other API.

I think @Bento007 and I talked about this previously; he may have had a compelling argument for `PUT`. I agree that `GET` makes more sense.

danieljhegeman commented:
> RE [DELETE /curation/collections/{collection_uuid}](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/delete_curation_collections__collection_uuid_)
> 
> What is the return status if the `collection_uuid` is invalid - for example, the resource does not exist or a public collection uuid is specified, but there's no revision in progress.

invalid: 401
no revision in progress: 404

danieljhegeman commented:
> `mean_genes_per_cell` should be included with datasets - for download scenarios rather than for curators.

Ok; the exact features of a dataset that can be sourced by a download link that we provide is not something that lends itself to specification in this API, per se, because we're just returning a link in the download scenario. Do you want (me) to document this somehow for the curators?

danieljhegeman commented:
> RE [DELETE /curation/collections/{collection_uuid}/datasets](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/delete_curation_collections__collection_uuid__datasets)
> 
> The documentation for the `curator_tag` and `dataset_uuid` parameters is unclear. Can you provide examples? If a curator does not pass either, what's the behavior?

Added documentation for every endpoint that requires one of those two parameters, and added examples for the entities themselves. 

If a curator does not pass either, we need to return a 400. Added in `b6e73ff5`.

danieljhegeman commented on backend/open_api_spec.yml:
```diff
@@ -0,0 +1,830 @@
+          $ref: "#/components/responses/403"
+        "404":
+          $ref: "#/components/responses/404"
+    patch:
```
Should updating the curator_tag require that the Collection not be public?

When just considering the data involved and whether it is "okay" to update tags, which are not surfaced to other users, it seems like it would be fine to all the tags for a public Collection to be modified. However, I'm not sure if allowing the `dataset.updated_at` to be greater/after the `project.revised_at` date for its Collection might cause logical errors elsewhere in the pipeline.

brianraymor replied:
Per the table in [The data model must be updated to support filtering by date ranges](https://github.com/chanzuckerberg/single-cell-data-portal/issues/1339), updating collection *information* does not update the timestamps in question:

| Operation | collection published_at | collection revised_at | dataset published_at | dataset revised_at |
|:---:|:---:|:---:|:---:|:---:|
| Publish private collection | X | - | X | - |
| Update a dataset in revision | -  | X | -  | X |
| Add a dataset in revision | - | X | X | - |
| Delete a dataset in revision | - | X | - | - |
| Update collection information in a revision | - | -  | - | - |

Tags would be a similar case.

brianraymor commented:
RE

> mean_genes_per_cell should be included with datasets - for download scenarios rather than for curators.

@danieljhegeman responded:

> Ok; the exact features of a dataset that can be sourced by a download link that we provide is not something that lends itself to specification in this API, per se, because we're just returning a link in the download scenario. Do you want (me) to document this somehow for the curators?

When I wrote `datasets`, I was referring to the dataset metadata returned in the API response. For example, here's a fragment from the current portal API:

        "id": "524e045e-e74c-4e00-9884-d5c3bef3d862",
        "is_primary_data": "SECONDARY",
        "is_valid": false,
        "mean_genes_per_cell": 1316.84206695779,
        "name": "Healthy human liver: hepatic stellate cells",

brianraymor commented:
> I'd like to add more details (sample code) for how to use the boto3 package to upload the datasets, 

And/or [PAWS](https://www.r-bloggers.com/2019/10/an-amazon-sdk-for-r/) ? We should be language neutral.

brianraymor commented:
RE [DELETE /curation/collections/{collection_uuid}/datasets](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/delete_curation_collections__collection_uuid__datasets)

The documentation states:


_Delete a Dataset. **The Dataset MUST NOT belong to a published Collection**. MUST indicate the Dataset by providing one of either the curator_tag (non-null) OR the dataset_uuid parameters._

But curators may delete a dataset that belongs to a published collection IF the operation occurs in a private revision of the public collection. 

How about:

_Delete a Dataset. The `collection_uuid` MUST reference either a private collection or a private revision of a public collection._

danieljhegeman commented:
> RE [DELETE /curation/collections/{collection_uuid}/datasets](https://app.swaggerhub.com/apis/danieljhegeman/Curation/1#/default/delete_curation_collections__collection_uuid__datasets)
> 
> The documentation states:
> 
> _Delete a Dataset. **The Dataset MUST NOT belong to a published Collection**. MUST indicate the Dataset by providing one of either the curator_tag (non-null) OR the dataset_uuid parameters._
> 
> But curators may delete a dataset that belongs to a published collection IF the operation occurs in a private revision of the public collection.
> 
> How about:
> 
> _Delete a Dataset. The `collection_uuid` MUST reference either a private collection or a private revision of a public collection._

Agreed, your wording is more clear. Changed in `18e6cb68`

danieljhegeman commented:
> PAWS

Yes! Stretch goal, likely.

At present I have a python gist that I can pass on to curators. I don't know how to get the links to work in Swagger, and from what I've read, it's not possible.

danieljhegeman commented:
> RE
> 
> > mean_genes_per_cell should be included with datasets - for download scenarios rather than for curators.
> 
> @danieljhegeman responded:
> 
> > Ok; the exact features of a dataset that can be sourced by a download link that we provide is not something that lends itself to specification in this API, per se, because we're just returning a link in the download scenario. Do you want (me) to document this somehow for the curators?
> 
> When I wrote `datasets`, I was referring to the dataset metadata returned in the API response. For example, here's a fragment from the current portal API:
> 
> ```
>     "id": "524e045e-e74c-4e00-9884-d5c3bef3d862",
>     "is_primary_data": "SECONDARY",
>     "is_valid": false,
>     "mean_genes_per_cell": 1316.84206695779,
>     "name": "Healthy human liver: hepatic stellate cells",
> ```

Got it, sorry for the confusion. 

Separately, I want to note that `mean_genes_per_cell` was not mentioned as a desired attribute in @jahilton 's [list](https://czi-sci.slack.com/archives/C02UGUV5R3J/p1644454526450299), but we can add it. Same for `is_valid`.

brianraymor commented:
Please add `mean_genes_per_cell` but not `is_valid` - which I assume has something to do with successful dataset validation.

brianraymor commented:
Please add `batch_condition` - per the comment in [**#curation-api**](https://czi-sci.slack.com/archives/C02UGUV5R3J/p1645831434417039).
